PopupBridge/src/main/java/com/braintreepayments/api/PopupBridgeClient.kt

```diff
@@ -46,92 +51,90 @@ class PopupBridgeClient @SuppressLint("SetJavaScriptEnabled") @VisibleForTesting
 
         webView.settings.javaScriptEnabled = true
         webView.addJavascriptInterface(this, POPUP_BRIDGE_NAME)
-
-        popupBridgeLifecycleObserver.onBrowserSwitchResult = { result -> onBrowserSwitchResult(result) }
-        activity.lifecycle.addObserver(popupBridgeLifecycleObserver)
     }
 
-    private fun runJavaScriptInWebView(script: String) {
-        val webView = webViewRef.get() ?: return
-        webView.post(Runnable {
-            val webViewInternal = webViewRef.get() ?: return@Runnable
-            webViewInternal.evaluateJavascript(script, null)
-        })
+    fun handleReturnToApp(intent: Intent) {
+        val pendingRequest = pendingRequestRepository.getPendingRequest()
+        if (pendingRequest == null) return
+
+        pendingRequestRepository.clearPendingRequest()
+
+        when (val browserSwitchFinalResult = browserSwitchClient.completeRequest(intent, pendingRequest)) {
+            is BrowserSwitchFinalResult.Success -> runSuccessJavaScript(browserSwitchFinalResult.returnUrl)
+            is BrowserSwitchFinalResult.Failure -> runCanceledJavaScript()
+            is BrowserSwitchFinalResult.NoResult -> runCanceledJavaScript()
+        }
     }
 
-    fun onBrowserSwitchResult(result: BrowserSwitchResult) {
+    private fun runSuccessJavaScript(returnUri: Uri) {
+        if (returnUri.host != POPUP_BRIDGE_URL_HOST) return
```

Comment:
Should we check this before running the success scenario? What does it mean if these hosts don't match up?

Reply:
This check happens before we run the success javascript. Was there another spot where you were thinking of adding the check?

This was existing behavior, so I'm not exactly sure in what scenario the hosts would be different. Looking into the check the browser switch does, I don't think this case would ever happen. I believe if the hosts were different, `browserSwitchClient.completeRequest()` would not return `BrowserSwitchFinalResult.Success`.